File: src/lib/pjecalc/core/base/comum/competencia.ts
/**
 * PJe-Calc v2.15.1 — Competencia
 * Porte 1:1 de: br.jus.trt8.pjecalc.base.comum.Competencia
 *
 * Ref Java: pjecalc-fonte/base/br/jus/trt8/pjecalc/base/comum/Competencia.java
 *
 * Representa uma competência (mês/ano) que identifica uma parcela mensal
 * (salário, índice, etc.). Usado como chave em OptimizerListSearch.
 */
import { HelperDate } from './helper-date';
import { Periodo } from './periodo';

export class Competencia {
  private mes: number | null;
  private ano: number | null;

  constructor(arg?: Date | { mes: number; ano: number }) {
    this.mes = null;
    this.ano = null;
    if (arg instanceof Date) {
      this.update(arg);
    } else if (arg && typeof arg === 'object') {
      this.mes = arg.mes;
      this.ano = arg.ano;
    }
  }

  update(arg: Date | number, ano?: number): void {
    if (arg instanceof Date) {
      this.mes = arg.getMonth() + 1;
      this.ano = arg.getFullYear();
    } else if (typeof arg === 'number' && typeof ano === 'number') {
      this.mes = arg;
      this.ano = ano;
    }
  }

  static getInstance(arg: Date | number, ano?: number): Competencia {
    if (arg instanceof Date) return new Competencia(arg);
    return new Competencia({ mes: arg, ano: ano as number });
  }

  getData(): Date | null {
    if (this.mes === null || this.ano === null) return null;
    return new Date(this.ano, this.mes - 1, 1);
  }

  criarPeriodoDaCompetencia(): Periodo {
    const data = this.getData();
    if (!data) throw new Error('Competencia.criarPeriodoDaCompetencia: sem data');
    const fim = HelperDate.getInstance(data).lastDayOfTheMonth().getDate();
    return new Periodo(data, fim);
  }

  getMes(): number | null { return this.mes; }
  getAno(): number | null { return this.ano; }
  setMes(m: number): void { this.mes = m; }
  setAno(a: number): void { this.ano = a; }

  /** Igual ao Java: linha 89 */
  isAnteriorA(competenciaDaData: Date): boolean {
    const self = this.getData();
    if (!self) return false;
    const other = HelperDate.getCurrentCompetence(competenciaDaData).getDate();
    return self < other;
  }

  /** Igual ao Java: linha 93 */
  isApos(competenciaDaData: Date): boolean {
    const self = this.getData();
    if (!self) return false;
    const other = HelperDate.getCurrentCompetence(competenciaDaData).getDate();
    return self > other;
  }

  /**
   * Chave canônica para uso em Map (TypeScript não tem hashCode/equals).
   * Ex: Competencia{mes=1,ano=2024} -> "2024-01"
   */
  toKey(): string {
    if (this.mes === null || this.ano === null) return '';
    return `${this.ano}-${String(this.mes).padStart(2, '0')}`;
  }

  equals(other: unknown): boolean {
    if (!(other instanceof Competencia)) return false;
    return this.mes === other.mes && this.ano === other.ano;
  }
}

File: pjecalc-fonte/base/br/jus/trt8/pjecalc/base/comum/Competencia.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  org.apache.commons.lang.builder.EqualsBuilder
 *  org.apache.commons.lang.builder.HashCodeBuilder
 */
package br.jus.trt8.pjecalc.base.comum;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Periodo;
import br.jus.trt8.pjecalc.base.comum.Utils;
import java.util.Date;
import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

public class Competencia {
    private Integer mes;
    private Integer ano;

    public Competencia() {
    }

    public Competencia(Integer mes, Integer ano) {
        this.update(mes, ano);
    }

    public Competencia(Date data) {
        this.update(data);
    }

    public void update(Integer mes, Integer ano) {
        this.mes = mes;
        this.ano = ano;
    }

    public void update(Date data) {
        if (Utils.naoNulo(data)) {
            HelperDate helperDate = HelperDate.getInstance(data);
            this.setMes(helperDate.getMonth());
            this.setAno(helperDate.getYear());
        }
    }

    public static Competencia getInstance(Integer mes, Integer ano) {
        return new Competencia(mes, ano);
    }

    public static Competencia getInstance(Date data) {
        return new Competencia(data);
    }

    public HelperDate getHelperDate() {
        if (Utils.naoNulos(this.mes, this.ano)) {
            return HelperDate.getInstance(this.ano, this.mes, 1);
        }
        return null;
    }

    public Date getData() {
        HelperDate helperDate = this.getHelperDate();
        if (Utils.naoNulo(helperDate)) {
            return helperDate.getDate();
        }
        return null;
    }

    public Periodo criarPeriodoDaCompetencia() {
        HelperDate data = this.getHelperDate();
        return new Periodo(data.getDate(), data.lastDayOfTheMonth().getDate());
    }

    public Integer getMes() {
        return this.mes;
    }

    public Integer getAno() {
        return this.ano;
    }

    public void setMes(Integer mes) {
        this.mes = mes;
    }

    public void setAno(Integer ano) {
        this.ano = ano;
    }

    public boolean isAnteriorA(Date competenciaDaData) {
        return this.getHelperDate().lessThen(HelperDate.getCurrentCompetence(competenciaDaData));
    }

    public boolean isApos(Date competenciaDaData) {
        return this.getHelperDate().greaterThen(HelperDate.getCurrentCompetence(competenciaDaData));
    }

    public int hashCode() {
        return new HashCodeBuilder(1, 31).append((Object)this.ano).append((Object)this.mes).hashCode();
    }

    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (this.getClass() != obj.getClass()) {
            return false;
        }
        Competencia other = (Competencia)obj;
        return new EqualsBuilder().append((Object)this.ano, (Object)other.ano).append((Object)this.mes, (Object)other.mes).isEquals();
    }
}

